akka-stream-typed/src/main/scala/akka/stream/typed/scaladsl/ActorFlow.scala

```diff
@@ -0,0 +1,126 @@
+   *
+   * '''Fails when''' the passed in actor terminates, or a timeout is exceeded in any of the asks performed
+   *
+   * '''Cancels when''' downstream cancels
+   *
+   * @tparam I Incoming element type of the Flow
+   * @tparam Q Question message type that is spoken by the target actor
+   * @tparam O answer type that the Actor is expected to reply with, it will become the Output type of this Flow
+   */
+  @implicitNotFound("Missing an implicit akka.util.Timeout for the ask() stage")
+  def ask[I, Q, O](ref: ActorRef[Q])(makeMessage: (I, ActorRef[O]) ⇒ Q)(implicit timeout: Timeout): Flow[I, O, NotUsed] =
+    ask(parallelism = 2)(ref)(makeMessage)(timeout)
+
+  /**
+   * Use the `ask` pattern to send a request-reply message to the target `ref` actor.
+   * If any of the asks times out it will fail the stream with a [[java.util.concurrent.TimeoutException]].
+   *
+   * Do not forget to include the expected response type in the method call, like so:
+   *
+   * {{{
+   * flow.via(ActorFlow.ask[String, Asking, Reply](parallelism = 4)(ref, (el, replyTo) => Asking(el, replyTo)))
+   *
+   * // or even:
+   * flow.via(ActorFlow.ask[String, Asking, Reply](parallelism = 4)(ref, Asking(_, _)))
+   * }}}
+   *
+   * otherwise `Nothing` will be assumed, which is most likely not what you want.
+   *
+   * The stage fails with an [[akka.stream.WatchedActorTerminatedException]] if the target actor is terminated,
+   * or with an [[java.util.concurrent.TimeoutException]] in case the ask exceeds the timeout passed in.
+   *
+   * Adheres to the [[ActorAttributes.SupervisionStrategy]] attribute.
+   *
+   * '''Emits when''' the futures (in submission order) created by the ask pattern internally are completed
+   *
+   * '''Backpressures when''' the number of futures reaches the configured parallelism and the downstream backpressures
+   *
+   * '''Completes when''' upstream completes and all futures have been completed and all elements have been emitted
+   *
+   * '''Fails when''' the passed in actor terminates, or a timeout is exceeded in any of the asks performed
+   *
+   * '''Cancels when''' downstream cancels
+   *
+   * @tparam I Incoming element type of the Flow
+   * @tparam Q Question message type that is spoken by the target actor
+   * @tparam O answer type that the Actor is expected to reply with, it will become the Output type of this Flow
+   */
+  @implicitNotFound("Missing an implicit akka.util.Timeout for the ask() stage")
+  def ask[I, Q, O](parallelism: Int)(ref: ActorRef[Q])(makeMessage: (I, ActorRef[O]) ⇒ Q)(implicit timeout: Timeout): Flow[I, O, NotUsed] = {
+    import akka.actor.typed.scaladsl.adapter._
+    val untypedRef = ref.toUntyped
+
+    val askFlow = Flow[I]
+      .watch(untypedRef)
+      .mapAsync(parallelism) { el ⇒
+        akka.pattern.extended.ask(untypedRef, (replyTo: akka.actor.ActorRef) ⇒ makeMessage(el, replyTo))
+      }
+      .map(_.asInstanceOf[O])
```

Comment:
in untyped we use a ClassTag, why there and not here?

there we also use `mapTo` of the `Future`, doesn't that give better error messages?

Reply:
ok, I was thinking of Java that the class is a bit of a hassle; but you're right that the API should be 1:1 with the existing one. Adding

Reply:
Hmm... actually it really would be nice to avoid the class parameter... since correctness is achieved by the types of actor refs, so would be nice go not have people to repeat this.
Without the cast it was unhappy, but I can do the cast on the Future, which would be a generic T there so no-op right?

Reply:
Proposing instead:

```
      .mapAsync(parallelism) { el ⇒
        val res = akka.pattern.extended.ask(untypedRef, (replyTo: akka.actor.ActorRef) ⇒ makeMessage(el, replyTo))
        // we need to cast manually (yet safely, by construction!) since otherwise we need a ClassTag,
        // which in Scala is fine, but then we would force JavaDSL to create one, which is a hassle in the Akka Typed DSL,
        // since one may say "but I already specified the type!", and that we have to go via the untyped ask is an implementation detail
        res.asInstanceOf[Future[A]]
      }
```

Reply:
ok, I was mostly interested in the motivation. 
cast of the Future or inside the map doesn't matter.

Reply:
👍